model/DataBaseAccess.php

```diff
@@ -65,6 +68,284 @@ protected function getEventManager()
+SQL;
+
+        return array_column(
+            $this->fetchQuery(
+                $query,
+                [
+                    $resourceUri,
+                    OntologyRdfs::RDFS_SUBCLASSOF,
+                    OntologyRdf::RDF_TYPE,
+                    OntologyRdfs::RDFS_SUBCLASSOF,
+                    OntologyRdf::RDF_TYPE,
+                    'http://www.tao.lu/Ontologies/TAO.rdf#AssessmentContentObject',
+                    'http://www.tao.lu/Ontologies/TAO.rdf#TAOObject',
+                    'http://www.tao.lu/Ontologies/generis.rdf#generis_Ressource',
+                    'http://www.w3.org/2000/01/rdf-schema#Resource',
+                ]
+            ),
+            'object'
+        );
+    }
+
+    public function getClassesResources(array $classesIds): array
+    {
+        $inQuery = implode(',', array_fill(0, count($classesIds), '?'));
+        $query = "SELECT subject, object FROM statements WHERE predicate IN (?, ?) AND object IN ($inQuery)";
+
+        $results = $this->fetchQuery(
+            $query,
+            [
+                OntologyRdfs::RDFS_SUBCLASSOF,
+                OntologyRdf::RDF_TYPE,
+                ...$classesIds,
+            ]
+        );
+
+        $tree = [];
+
+        foreach ($classesIds as $classId) {
+            $resources = array_filter($results, static fn (array $result): bool => $result['object'] === $classId);
+            $tree[$classId] = array_column($resources, 'subject');
+        }
+
+        return $tree;
+    }
+
+    public function getResourceTree(core_kernel_classes_Resource $resource): array
+    {
+        $query = <<<'SQL'
+WITH RECURSIVE statements_tree AS (
+    SELECT
+        r.subject,
+        r.predicate,
+        1 as level
+    FROM statements r
+    WHERE r.subject = ?
+      AND r.predicate IN (?, ?)
+    UNION ALL
+    SELECT
+        s.subject,
+        s.predicate,
+        level + 1
+    FROM statements s
+        JOIN statements_tree st
+            ON s.object = st.subject
+)
+SELECT subject, predicate, level FROM statements_tree;
+SQL;
+
+        return $this->fetchQuery(
+            $query,
+            [
+                $resource->getUri(),
+                OntologyRdfs::RDFS_SUBCLASSOF,
+                OntologyRdf::RDF_TYPE,
+            ]
+        );
+    }
+
+    public function getPermissionsByUsersAndResources(array $userIds, array $resourceIds): array
+    {
+        // Permissions for an empty set of resources must be an empty array
+        if (empty($resourceIds)) {
+            return [];
+        }
+
+        $inQueryResources = implode(',', array_fill(0, count($resourceIds), '?'));
+        $inQueryUsers = implode(',', array_fill(0, count($userIds), '?'));
+
+        $query = <<<SQL
+SELECT
+    resource_id,
+    user_id,
+    privilege
+FROM data_privileges
+WHERE resource_id IN ($inQueryResources)
+  AND user_id IN ($inQueryUsers)
+SQL;
+
+        $results = $this->fetchQuery($query, [...$resourceIds, ...$userIds]);
```

Comment:
- [ ] We need to watch out the query size here. Depending on the amount of users and resources select the query might exceed its capacity

Reply:
Already done